### PR TITLE
Feature/authentication

### PR DESCRIPTION
**Pull request by ArtsiomSt**
Implemented authentication using JWT tokens

### CODE REVIEW

File: CarShowRoom/CarShowRoom/urls.py [hunk 1]
```diff
@@ -25,13 +26,24 @@
         path("sellers/", include("sellers.api.v1.urls")),
         path("cars/", include("cars.api.v1.urls")),
         path("customers/", include("customers.api.v1.urls")),
+        path("token/", TokenObtainPairView.as_view(), name="token_obtain_pair"),
+        path("token/refresh/", TokenRefreshView.as_view(), name="token_refresh"),
```
Comment (rupeq) [inline CarShowRoom/CarShowRoom/urls.py]: I don't think you should have it here. Its enough to have `token_urls`

File: CarShowRoom/core/serializers.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+class RegisterSerializer(serializers.ModelSerializer):
+    """This serializer implements registration for models, that need field balance"""
+
+    balance = BalanceSerializer(read_only=True)
+    password = serializers.CharField(
+        max_length=100, write_only=True, validators=[validate_password]
+    )
+
+    def create(self, validated_data):
+        new_balance = Balance.objects.create(money_amount=0)
```
Comment (rupeq) [inline CarShowRoom/core/serializers.py]: another approach here is to use signals in order to create balance once user is created

**Review by rupeq — COMMENTED** [review]
1 nit

File: CarShowRoom/CarShowRoom/settings.py [hunk 3]
```diff
@@ -142,8 +143,74 @@
...
+
+
+EMAIL_HOST = "smtp.yandex.ru"
+EMAIL_PORT = 465
+EMAIL_USE_TLS = False
+EMAIL_USE_SSL = True
+
+EMAIL_HOST_USER = "akeonst@yandex.ru"
+EMAIL_HOST_PASSWORD = os.getenv("EMAIl_HOST_PASSWORD")
+DEFAULT_FROM_EMAIL = "akeonst@yandex.ru"
```
Comment (rupeq) [inline CarShowRoom/CarShowRoom/settings.py]: Let's place email to env as well

File: CarShowRoom/customers/apps.py [hunk 4]
```diff
@@ -4,3 +4,6 @@
 class CustomersConfig(AppConfig):
     default_auto_field = "django.db.models.BigAutoField"
     name = "customers"
+
+    def ready(self):
+        import customers.signals.handlers
```
Comment (rupeq) [inline CarShowRoom/customers/apps.py:9]: Why do you need this?

File: CarShowRoom/CarShowRoom/settings.py [hunk 5]
```diff
@@ -49,9 +50,9 @@
     "core",
     "drf_yasg",
     "rest_framework",
-    "sellers",
-    "customers",
-    "cars",
+    "sellers.apps.SellersConfig",
+    "customers.apps.CustomersConfig",
+    "cars.apps.CarsConfig",
```
Comment (rupeq) [inline CarShowRoom/CarShowRoom/settings.py:55]: Why?

File: CarShowRoom/CarShowRoom/settings.py [hunk 6]
```diff
@@ -142,8 +143,74 @@
...
+    "default": {
+        "BACKEND": "django_redis.cache.RedisCache",
+        "LOCATION": "redis://{}:{}/1".format(os.getenv("REDIS_HOST", "redis"), os.getenv("REDIS_PORT", "6379")),
+        "OPTIONS": {
+            "CLIENT_CLASS": "django_redis.client.DefaultClient",
+        },
+    }
+}
+
+USER_CONFIRMATION_KEY = "user_confirmation_{token}"
```
Comment (rupeq) [inline CarShowRoom/CarShowRoom/settings.py]: Let's place it in env

File: CarShowRoom/core/models.py [hunk 7]
```diff
@@ -24,3 +24,11 @@ class CarPriceCurrency(models.Model):
 
     class Meta:
         abstract = True
+
+
+class User(AbstractUser):
```
Comment (rupeq) [inline CarShowRoom/core/models.py:29]: Let's add required fields = username and username field = email

File: CarShowRoom/core/mixins.py [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+            request = self.context["request"]
+        confirm_link = request.build_absolute_uri(
+            reverse_lazy("tokens:email_confirm", kwargs={"token": token})
+        )
+        send_mail(
+            topic,
+            f"{message_before_link}\n{confirm_link}",
+            "akeonst@yandex.ru",
+            [instance.email],
+        )
```
Comment (rupeq) [inline CarShowRoom/core/mixins.py]: I don't think it should be a mixin. You can create service.py and place function there not a class. And then import function where you need

File: CarShowRoom/core/mixins.py [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+        if request is None:
+            request = self.context["request"]
+        confirm_link = request.build_absolute_uri(
+            reverse_lazy("tokens:email_confirm", kwargs={"token": token})
+        )
+        send_mail(
+            topic,
+            f"{message_before_link}\n{confirm_link}",
+            "akeonst@yandex.ru",
```
Comment (rupeq) [inline CarShowRoom/core/mixins.py]: Env?

File: CarShowRoom/customers/signals/handlers.py [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+
+from customers.models import Balance, Customer
+
+
+@receiver(post_save, sender=Customer)
+def link_balance(sender, instance, **kwargs):
+    if kwargs.get("created", False):
+        new_balance = Balance.objects.create(money_amount=0)
+        setattr(instance, "balance", new_balance)
+        instance.save()
```
Comment (rupeq) [inline CarShowRoom/customers/signals/handlers.py]: Can you create just one signal for user model since every model inherits from user ?

**Review by rupeq — CHANGES_REQUESTED** [review]

File: CarShowRoom/customers/apps.py (hunk 4 above)
Comment (ArtsiomSt) [reply CarShowRoom/customers/apps.py:9]: It is used to make signals work, they didnt work, and i found that solutions on stackoverflow)

File: CarShowRoom/CarShowRoom/settings.py (hunk 5 above)
Comment (ArtsiomSt) [reply CarShowRoom/CarShowRoom/settings.py:55]: It is used to make signals work, they didnt work, and i found that solutions on stackoverflow)

File: CarShowRoom/core/mixins.py (hunk 8 above)
Comment (ArtsiomSt) [reply CarShowRoom/core/mixins.py]: Agree, will change

File: CarShowRoom/core/mixins.py (hunk 9 above)
Comment (ArtsiomSt) [reply CarShowRoom/core/mixins.py]: What do you mean?

File: CarShowRoom/customers/signals/handlers.py (hunk 10 above)
Comment (ArtsiomSt) [reply CarShowRoom/customers/signals/handlers.py]: Agree, will try to change its behavior, but i am afraid of admin-users behavior, so i will check it

File: CarShowRoom/core/mixins.py (hunk 9 above)
Comment (rupeq) [reply CarShowRoom/core/mixins.py]: Place email to env

File: docker-compose.yml [hunk 11]
```diff
@@ -21,6 +21,11 @@ services:
       - env/db.env
     ports:
       - "5555:5432"
+  redis:
+    image: redis
+    container_name: redis
+    ports:
+       - "127.0.0.1:16379:6379"
```
Comment (rupeq) [inline docker-compose.yml]: Let's change it. 
To env (for backend) add `REDIS_HOST=redis://redis:6379`
Change here image redis:alpine, host name "redis", restart always

Don't open ip address as now in ports

File: CarShowRoom/customers/apps.py (hunk 4 above)
Comment (ArtsiomSt) [reply CarShowRoom/customers/apps.py:9]: They didnt work when they were located in separate file, so i used this tip

File: CarShowRoom/core/mixins.py (hunk 9 above)
Comment (ArtsiomSt) [reply CarShowRoom/core/mixins.py]: Oh, okey)

File: CarShowRoom/customers/signals/handlers.py (hunk 10 above)
Comment (ArtsiomSt) [reply CarShowRoom/customers/signals/handlers.py]: I have checked it, when we use model _User_ as sender it does not catch signal from inherited models

Comment (rupeq) [reply CarShowRoom/customers/signals/handlers.py]: you can try to use init subclasses dunder for User model:

```
    @classmethod
    def __init_subclass__(cls, **kwargs):
        super().__init_subclass__(**kwargs)
        models.signals.post_save.connect(perform_on_save, sender=cls)
```

Comment (rupeq) [reply CarShowRoom/customers/signals/handlers.py]: btw thats okey what you already did

File: docker-compose.yml [hunk 12]
```diff
@@ -22,10 +22,12 @@ services:
     ports:
       - "5555:5432"
   redis:
-    image: redis
+    image: redis:alpine
     container_name: redis
+    hostname: redis
     ports:
-       - "127.0.0.1:16379:6379"
+       - "16379:6379"
```
Comment (rupeq) [inline docker-compose.yml]: why do you need outer port `16379`?

Comment (ArtsiomSt) [reply docker-compose.yml]: There was an exception that port  6379 is already in use, so i used that one

File: CarShowRoom/CarShowRoom/settings.py [hunk 13]
```diff
@@ -196,21 +196,21 @@
 EMAIL_USE_TLS = False
 EMAIL_USE_SSL = True
 
-EMAIL_HOST_USER = "akeonst@yandex.ru"
+EMAIL_HOST_USER = os.getenv("EMAIL_USER")
 EMAIL_HOST_PASSWORD = os.getenv("EMAIl_HOST_PASSWORD")
```
Comment (rupeq) [inline CarShowRoom/CarShowRoom/settings.py]: nit: typo (lower case)

File: docker-compose.yml (hunk 12 above)
Comment (rupeq) [reply docker-compose.yml]: try to expose 6379 port and then try again

**Review by rupeq — APPROVED** [review]

**Pull request merged**